### DISCUSSION TITLE
Concept URI vs dereferenced web page URI

### DISCUSSION THREAD

**Discussion by ecow** [edmcouncil/fibo#1363] [General] [answered]
Hi,
I'm developing semantic web application using some  FIBO concepts and I have a question about URI dereferencing in FIBO:

It seems to me that you are using the same URI for the web page (i.e. informative resource) and for the abstract concept (i.e. non-informative resource). If this is  true it could lead to a semantic collision according with the Web architecture specifications (cfg. https://www.w3.org/TR/webarch/#URI-collision and https://www.w3.org/TR/cooluris/#distinguishing). I'm wrong? How do you manage this?

e.g.:

is the dct:title a property of  the https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney  
URI subject  referring to the web page title or of the  fibo-fnd-acc-cur:AmountOfMoney conceptual owl:Class?

How can I disambiguate the two things?

Many thanks for your support

Best regards

Enrico Fagnoni

**Comment by trypuz** [comment]
Hi Enrico (@ecow)! We are not using the URI to mean two different things. We dereference URIs using content negotiation mechanisms to provide different views on URIs for different agents.

**Comment by ecow** [comment]
I do not understand. How you differentiate the web Page from the FIBO concept?

In other words if you say:

<https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney> dcterm:created "20200701"    .

Are you referring the date of creation of  the FIBO concept or the date of the creation of the Page describing it?

To use the same uri for two different meanings (a web Page is a foaf:Document) breaks web standards.

IMHO, this has nothing to do with content negotiation.

**Comment by trypuz** [comment]
Everything in the FIBO code applies to the ontological resources (concepts, properties, etc.). It has nothing to do with the FIBO Viewer and the web pages.

**Comment by dallemang** [comment]
I'm a bit confused by this discussion - it seem to have started with the observation "[EDMC] are using the same URI for the web page ... and for the abstract concept ..."   I'm not sure what this observation is based on; if somewhere we have a triple that says something like: 

<https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney>  a  ex:WebPage .

then it is indeed an error.  I haven't been able to find any such triple. 

If it has to do with the content negotiation (despite Enrico's comment that it doesn't), that is, the fact that if you resolve https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney with a web browser, you get a web page, then this seems indeed to be mistaking content negotiation for a statement of type.    It is pretty common practice (and many FIBO users for years scolded us for not doing this) to have RDF resources resolve to web pages about themselves when resolved on behalf of a web browser.  For example, the SKOS URL http://www.w3.org/2004/02/skos/core# resolves to a web page about the standard when resolved using a web browser, and resolves to an RDF/XML file when resolved using an appropriate media type (e.g., application/rdf+xml).     Similar comments apply to Dublin Core terms (http://purl.org/dc/terms/) and XML Schema ( http://www.w3.org/2001/XMLSchema#)

FIBO resources resolve to RDF fragments in appropriate serializations when fetched with corresponding media types as outlined in https://www.w3.org/2008/01/rdf-media-types; For example, 

curl -H "Accept: application/rdf+xml"  https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney

returns triples in RDF/XML format


We have done our best to conform to W3C practice, by imitating what they themselves have done for similar efforts (e.g., SKOS, XSD) and what other have done in similar contexts (e.g., Dublin core).  We have also responded to complaints by others that by failing to resolve FIBO resources to human readable pages when resolved on behalf of a web browser, we were out of compliance with best web practices.  If there is some way in which we have fallen short, please help us to understand what we should do differently.  But I don't see how we have made a statement that AmountOfMoney is a web page any more than the W3C has made a statement that <http://www.w3.org/2004/02/skos/core#> is a web page.  They explicitly state that 

<http://www.w3.org/2004/02/skos/core#> a owl:Ontology .

just as we explicitly state 

fibo-fnd-acc-cur:AmountOfMoney  a owl:Class .

**Comment by ecow** [comment]
I understand,  But anyway you publish web pages that use the same URI. To be consistent with web standard you must use a different URI to represent the FIBO concept and the related  FIBO Viewer web page.

for instance:
- `https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney` is the FIBO concept
- `https://doc.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney` is the FIBO viewer  page describing the FIBO concept.

Another viable solution could be to use hash notation, i.e.:

-`https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney#doc` for distinguish FIBO viewer  page 

or service notation:

`https://spec.edmcouncil.org/fibo/view?uri=https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney`

As an example, the dpedia.org choice was to use:

- `http://dbpedia.org/resource/Milan`  is about the city 
- `http://dbpedia.org/page/Milan` is the web resource (i.e. the page) that contains the descriptions about the  http://dbpedia.org/resource/Milan concept (that also supports http  content negotiation to expose RDF data about )

Dereferencing without generating collision is one of the main requirement in the  Web and Linked Data standards. For more info see https://www.w3.org/TR/webarch/#URI-collision

I think this question should be moved to an issue.

Thank you for your support

**Comment by ecow** [comment]
Ps @dallemang  look well:

`<http://www.w3.org/2004/02/skos/core>  a owl:Ontology`    (wihout hash)

and it is dereferenced to `https://www.w3.org/2009/08/skos-reference/skos.html` or `http://www.w3.org/TR/skos-reference/skos.rdf` depending from content negotiation.

<http://www.w3.org/2004/02/skos/core>  is an ontology (i.we. a non informative web resource), `https://www.w3.org/2009/08/skos-reference/skos.(html|rdf)`  is an informative web resource

It seams a confusion between "dereferencing"  and "content negotiation", they are different things that address different problems in web architecture and in http protocol specification.



<http://www.w3.org/2004/02/skos/core#> is a name space without any formal meaning (for what i understand)

**Comment by trypuz** [comment]
@ecow, thank you for your comments. Of course, you can use the service notation (URLs with a query string) to refer to the FIBO web pages (e.g.:
 https://spec.edmcouncil.org/fibo/ontology?query=https%3A%2F%2Fspec.edmcouncil.org%2Ffibo%2Fontology%2FSEC%2FDebt%2FBonds%2F)

I asked @przemekgradzki to analyze the situation. We will soon find out what his opinion is.

**Comment by przemekgradzki** [comment]
@ecow , @dallemang , @trypuz: what do you think about this solution:
- `https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney` := identifier (non informative web resource)
- `https://spec.edmcouncil.org/fibo/ontology/viewer/FND/Accounting/CurrencyAmount/AmountOfMoney` := web page (informative web resource)
- `https://spec.edmcouncil.org/fibo/ontology/master/latest/FND/Accounting/CurrencyAmount/AmountOfMoney` := RDF document (informative web resource)

I think, that correct URI for "RDF document" is:
`https://spec.edmcouncil.org/fibo/ontology/master/latest/FND/Accounting/CurrencyAmount/`

**Comment by dallemang** [comment]
I like your first and third URI. 

One might consider, instead of 

https://spec.edmcouncil.org/fibo/ontology/viewer/FND/Accounting/CurrencyAmount/AmountOfMoney


to use

https://spec.edmcouncil.org/fibo/viewer/FND/Accounting/CurrencyAmount/AmountOfMoney

since this entity is not really in the ontology at all; it is a human-viewable entity.   As analogy, the URIs is the Vocabulary have names like .../fibo/vocabulary/...   and in the data dictionary .../fibo/datadictionary/... , that is, the product name comes right after fibo.   Since the viewer is also a product, wouldn't it make sense to put its name at that location?

**Comment by przemekgradzki** [comment]
@dallemang good point

**Comment by ecow** [comment]
Yea, all proposed solution works well.

My suggestion is to use the [Recipe 5 in the W3C document  Best Practice Recipes for Publishing RDF Vocabularies](https://www.w3.org/TR/swbp-vocab-pub/#recipe5)

so, the `spec.edmcouncil.org` server could be configured to redirect (303)  all URIs like `https://spec.edmcouncil.org/fibo/ontology/(.*)` to
`https://spec.edmcouncil.org/fibo/doc/$1`

than `https://spec.edmcouncil.org/fibo/doc/{.*}` could be a an API endpoint able to serve directly  the informative resources with content negotiation. 

The only unavoidable drawback is that someone could be tempted to cut&and paste the link in the the browser  URL bar and try to use it  instead of the concept URI. But this risk is already well mitigated by the presence of the [COPY] button in the viewer.

This is a [well known topic](https://en.wikipedia.org/wiki/The_Treachery_of_Images) :-)

**Comment by CoryCasanave** [comment]
I have always thought it very odd that any “pure” identifier would include a technical protocol (http:/https:) rather than something to indicate that it is just identity without expectation of access, like “id:”.
Where an access protocol is specified, it seems reasonable for that identifier to indicate where information about that identity can be found, at least information published by the authority minting the URI. It does not need to mean that the URI is the identity of that referenced resource and the thing, just that the identity of the thing can be dereferenced to information about the thing. Without any link between identity and information we would be lost. Pure identity is nice in theory but impractical in any open system.

-Cory

**Comment by rivettp** [comment]
I agree with @trypuz 's initial reaction, @dallemang 's line of questioning, and @CoryCasanave's last remark - though I'm not so bothered at the inclusion of protocol in the URI. ..

IMO with FIBO the situation is not the equivalent of Milan: we don't have a separately curated web page about the concept - the page is just a query of all the things we know about the concept. FIBO is all concepts, it's not about managing web pages, as Dean said.
To take the example of "AmountOfMoney" there is no such thing as a web page describing it, except in a trivial sense, since the browser page is instantiated on the fly. You literally, by design, have no means to express anything useful about the page. In other words, in the language of https://www.w3.org/TR/webarch/#URI-collision, there is only one resource - the concept.

**Comment by trypuz** [comment]
The question is: does the current setting (including FIBO Viewer config, Nginx config, and FIBO) break any W3C/RFC/SW standard?

**Comment by przemekgradzki** [comment]
@trypuz [3. URIs for Real-World Objects](https://www.w3.org/TR/cooluris/#semweb):

```
2. Be unambiguous.
There should be no confusion between identifiers for Web documents and identifiers for other resources. URIs are meant to identify only one of them, so one URI can't stand for both a Web document and a real-world object.
```

The most important is: _...one URI **can't stand** for both a Web document and a real-world object..._

Now URI=`https://spec.edmcouncil.org/fibo/ontology/viewer/FND/Accounting/CurrencyAmount/AmountOfMoney` stands for:

1. identifier (non informative web resource)
2. web page (informative web resource) see `curl -H "Accept: text/html" https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney`
3. RDF document (informative web resource) see `curl -H "Accept: application/rdf+xml" https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney`

I think it's acceptable to have one URI for web page (2.) and RDF document (3.) - they are distinguished
using [content negotiation](https://www.w3.org/TR/swbp-vocab-pub/#negotiation), but not for identifier (1.)

**Comment by mereolog** [comment]
@przemekgradzki 
I don't think that the response to a curl request is the object that the request's url identifies or if it is, then no url can identify any real-world object. 

Suppose you have a URL that identifies a John Doe, a real person. No curl request will return John Doe himself, but only a document or file that somehow names or identifies him. But this cannot be raised as an argument that the URL in question identifies both John Doe and some file/document about him. All URLs direct to web resources and not to real world objects (well at least not until we end up in the Matrix), even if some of them identify real world objects. The fact that a given curl request returns an html document, is not relevant for the question what this request's url identifies.

In the case of FIBO the situation is more convoluted because FIBO consists of concepts and these are not like persons. Still, I think my point above stands also for FIBO as these concepts are probably not just parts of rdf/xml documents.

**Comment by przemekgradzki** [comment]
@mereolog 
The discussion does not apply to items returned by curl.

1. [Cool URIs for the Semantic Web](https://www.w3.org/TR/cooluris/) **is** W3C standard.
2. Part of [Cool URIs for the Semantic Web](https://www.w3.org/TR/cooluris/) ([3. URIs for Real-World Objects](https://www.w3.org/TR/cooluris/#semweb), point _2. Be unambiguous._) states

_...one URI can't stand for both a Web document and a real-world object..._
   
3. URI=`https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney` stands for Web document (web page or RDF document) **and** a real-world object

And _stands for_ does not mean **_is_**

**Comment by dallemang** [comment]
I am confused by point (3); this URI has "/viewer/" in it, and therefore stands for the web document (which, as @rivettp pointed out, does not even exist).  The URI for the concept of AmountOfMoney is https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney

(notice that /viewer/ is missing)

**Comment by przemekgradzki** [comment]
@dallemang - typo in URI 3.; fixed

**Comment by mereolog** [comment]
@przemekgradzki 
I think I share @rivettp view that current FIBO URIs do not stand for/identify web pages but they stand for/identify FIBO concepts.
Note also that https://www.w3.org/TR/cooluris/#semweb is just a (working) group note and not a recommendation while https://www.w3.org/standards/faq#std states that (only?) recommendations are W3C standards.

**Comment by przemekgradzki** [comment]
- @mereolog  I agree, that [https://www.w3.org/TR/cooluris/#semweb](https://www.w3.org/TR/cooluris/#semweb) **is not** W3C standard.
- [RFC7231, section 2.](https://tools.ietf.org/html/rfc7231#section-2)
> Each resource is identified by a Uniform Resource Identifier (URI)

[RFC7231, section 6.3.1.](https://tools.ietf.org/html/rfc7231#section-6.3.1):

> The payload sent in a 200 response depends on the request method.
For the methods defined by this specification, the intended meaning
of the payload can be summarized as:
GET  a representation of the target resource;

So URI identifies resource and using that URI you can get "a representation of the target resource" = web page.

- @CoryCasanave I agree that ..
> ... Without any link between identity and information we would be lost. Pure identity is nice in theory but impractical in any open system.

[Cool URIs for the Semantic Web](https://www.w3.org/TR/cooluris/) is a practical and simple method ([RFC7231, section 6.4.4. _303 See Other_ redirect](https://tools.ietf.org/html/rfc7231#section-6.4.4)) to link "pure identifiers" of FIBO concepts with URIs, which identifies web pages. But it's just proposal

**Comment by jgeluk** [comment]
We created this IRI scheme for FIBO years ago based on a number of ideas, one of which was the "follow your nose principle" where a user (usually a data modeler, SME or ontologist) would have to be able to click on an ontology IRI or a property IRI and, depending on the client-app from which he/she makes the click and the preferred MIME type of that client-app, would end up with the right content. If you dereference from Protege or Topbraid Composer you'd end up fetching RDF. If you dereference from a browser you'd end up HTML.

Multiple "serialization formats" of the same information. 

And yes that is a slight deviation from the linked data standard, I agree. Done on purpose at that time (by me).

One alternative at that time was to redirect in the frontend NGINX server, if you come in with a GET and an Accept header of text/html (like a browser would do) then redirect to the same IRI + `.html` at the end. That would have been more LD compliant. But would have other complications elsewhere.

**Comment by jgeluk** [comment]
@przemekgradzki about your proposal:

- `https://spec.edmcouncil.org/fibo/ontology/FND/Accounting/CurrencyAmount/AmountOfMoney` := identifier (non informative web resource)
- `https://spec.edmcouncil.org/fibo/ontology/viewer/FND/Accounting/CurrencyAmount/AmountOfMoney` := web page (informative web resource)
`https://spec.edmcouncil.org/fibo/ontology/master/latest/FND/Accounting/CurrencyAmount/AmountOfMoney` := RDF document (informative web resource)

Here's the original documentation of the "IRI scheme": https://github.com/edmcouncil/fibo/blob/master/etc/process/iri-scheme.md

The word `/ontology/` is the `<product>` name. You could consider "the viewer" to be just one of the many products that EDMC publishes for "product-family fibo". So you could use `https://spec.edmcouncil.org/fibo/ontology-viewer/FND/blah` and even support the versioned ontology IRIs then as well: `https://spec.edmcouncil.org/fibo/ontology-viewer/master/latest/FND/blah`

**Comment by jgeluk** [comment]
That would then also allow anyone else to come up with other types of "viewers" (maybe an editor even as some point? Or a collaborative ontology environment or some product of some partner?)

**Comment by przemekgradzki** [comment]
Let's call the new product `ontoviewer` and:

- `https://spec.edmcouncil.org/fibo/ontology/FND/...` would be the _ontology/property IRI_
- `https://spec.edmcouncil.org/fibo/ontoviewer/FND/...` := would be the _web page URL_ describing `https://spec.edmcouncil.org/fibo/ontology/FND/...` ontology/property
- `https://spec.edmcouncil.org/fibo/ontology/master/latest/FND/...` would be the _RDF document URL_ describing `https://spec.edmcouncil.org/fibo/ontology/FND/...` ontology/property

This is also in line with what @dallemang suggested in [this comment](https://github.com/edmcouncil/fibo/issues/1072#issuecomment-654898763)

If you use ontology/property IRI in your web browser you will be redirected to the web page - you will see the web page URL in the address bar

Is it acceptable?

**Comment by rivettp** [comment]
No I don't like the "ontoviewer" in the URI.
- as I said earlier, for FIBO the web page is not a distinct managed and editable resource - it's the output from querying information about the FIBO concept.
- the info/layout displayed by ontoviewer might change dynamically based on the user and other factors. e.g. business users might see less technical OWL information.
- users cannot really save or communicate the URI they get back - since it represents something ephemeral; and of course the tool deployed might change.
- still not sure what pragmatic problem this is trying to solve? In other words what would be the downside of closing this issue with no change? "If it ain't broke don't fix it"

**Comment by VladimirAlexiev** [comment]
Hi folks! Please use http://linkeddata.uriburner.com:8000/vapour to validate that resolution is done correctly, as listed at http://vocab.getty.edu/doc/#Semantic_Resolution.

@ecow thanks for bringing this up!
(btw, you mean "information vs non-information resource", not "informative": we hope that all our resources are informative :-))

**Comment by trypuz** [comment]
http://linkeddata.uriburner.com:8000/vapour?uri=https%3A%2F%2Fspec.edmcouncil.org%2Ffibo%2Fontology%2FBE%2FLegalEntities%2FLegalPersons%2FLegalEntity&defaultResponse=dontmind&userAgent=http%3A%2F%2Flinkeddata.uriburner.com%3A8000%2Fvapour%23this

Looks OK. Do we close this issue?

**Comment by ecow** [comment, marked as answer]
Sorry, I'm not so sure that it looks good...I read:"Conclusions on the type
of the resources

   - <
   https://spec.edmcouncil.org/fibo/ontology/BE/LegalEntities/LegalPersons/LegalEntity>
   identifies a Web document

"

This states that the  LegalEntity URI is a Web document. I do not think it
is what FIBO wants to assert.

Il mer 18 nov 2020, 11:59 Robert Trypuz <notifications@github.com> ha
scritto:
